File: train.js
//J-MITASK (NodeJS)
/*
Shunday function yozing, u parametridagi array ichida eng kop takrorlangan raqamni topib qaytarsin.
MASALAN: majorityElement([1,2,3,4,5,4,3,4]) return 4
 */
const majorityElement = (array) => {
    const counts = {};

    let maxCount = 0;
    let mostRepeated;

    for (let num of array) {
        counts[num] = (counts[num] || 0) + 1;

        if (counts[num] > maxCount) {
            maxCount = counts[num];
            mostRepeated = num
        }
    }
    return mostRepeated
}

const result = majorityElement([1,2,3,4,5,4,3,4])
console.log(result)

// [1,2,3,4,5,4,3,4]
// [1,2,3,4,5,4,3,4]


//H-TASK (NodeJS)
/*
shunday function tuzing, u integerlardan iborat arrayni argument sifatida qabul qilib, 
faqat positive qiymatlarni olib string holatda return qilsin
MASALAN: getPositive([1, -4, 2]) return qiladi "12"
*/
// const getPositive = (array) => {
//     newArray = []
//     for (let i = 0; i < array.length; i++) {
//         if (array[i] >= 0) {
//             newArray.push(array[i])
//         }
//     }
//     return newArray.join("")
// }

// const result = getPositive([1, -4, 2])
// console.log(result);




// MITASK-F
/* 
Shunday findDoublers function tuzing, unga faqat bitta string argument pass bolib, 
agar stringda bir hil harf qatnashgan bolsa true, qatnashmasa false qaytarishi kerak.
MASALAN: findDoublers("hello") return true return qiladi
*/

// h e l l o => h => e l l o=> l == h, e, l, o 

// const findDoublers = (string) => {
//     const setString = new Set(string)
//     console.log(setString, typeof setString)
//     const newString = [...setString].join('')
//     console.log(newString, typeof newString)
//     if (string.length !== newString.length) {
//         return true
//     } return false
// }

// const result = findDoublers('helo')
// console.log(result)



//MITASK-E
/*E-TASK (NodeJS)
Shunday function tuzing, u bitta string argumentni qabul 
qilib osha stringni teskari qilib return qilsin.
MASALAN: getReverse("hello") return qilsin "olleh" */
//SOLUTION
//DEFINE
// const getReverse = (word) => {
//     let newList = word.split("")
//     let reversedList = newList.reverse()
//     return reversedList.join("")
// }

// //CALL
// const result = getReverse('world')
// console.log(result)

//MITASK-D
/*Shunday function tuzingki unga integerlardan iborat array pass bolsin va 
function bizga osha arrayning eng katta qiymatiga tegishli birinchi indexni qaytarsin.
MASALAN: getHighestIndex([5, 21, 12, 21, 8]) return qiladi 1 sonini. */
//SOLUTION
//DEFINE
// const getHighestIndex = (array) => {
//     let highestEle = 0
//     for (let i = 0; i < array.length; i++) {
//         if (array[i] >= highestEle) {
//             highestEle = array[i]
//         }
//     }
//     return array.indexOf(highestEle)
// }


// //CALL
// const result = getHighestIndex([5, 6, 12, 21, 8, 42, 1, 56])
// console.log(result)


//MITASK-C
/*Shunday function tuzing, u 2ta string parametr ega bolsin, 
hamda agar har ikkala string bir hil harflardan iborat bolsa true aks holda false qaytarsin
MASALAN checkContent("mitgroup", "gmtiprou") return qiladi true;*/
//SOLUTION
//DEFINE
// const checkContent = (string1, string2) => {
//     if (typeof string1 != 'string' || typeof string2 != 'string') {
//         return "Please enter string arguments"
//     }

//     if (string1.length != string2.length) {
//         return false
//     }

//     let letter = ""
//     for (const char of string1) {
//         for (const ele of string2) {
//             if (char === ele) {
//                 letter += char
//             }
//         }
//     }
//     return true
// }

// //CALL
// const mitaskc = checkContent("muhammad", "mmmdahau")
// console.log(mitaskc)




//MITASK-B
/*Shunday function tuzing, u 1ta string parametrga ega bolsin, 
hamda osha stringda qatnashgan raqamlarni sonini bizga return qilsin.
MASALAN countDigits("ad2a54y79wet0sfgb9") 7ni return qiladi.*/
//SOLUTION
//DEFINE
// const numberCount = (numtext) => {
//     let count = 0;
//     for (const char of numtext) {
//         if (Number(char) == char) {
//             count += 1;
//         }
//     }
//     return count
// }

// //CALL
// const result = numberCount('ad2a54y79wet0sfgb9bdj7df890')
// console.log(result)



// MITASK-A
/*
Shunday 2 parametrli function tuzing, hamda birinchi parametrdagi 
letterni ikkinchi parametrdagi sozdan qatnashga sonini return qilishi kerak boladi.
MASALAN countLetter("e", "engineer") 3ni return qiladi.
*/


//SOLUTIION
// DEFINE
/*
const letterCount = (letter, word) => {
    if (typeof letter !== "string" || letter.length !== 1) {
        console.log("Please enter a single letter as a first argument")
    }

    if (typeof word !== 'string') {
        console.log("Please enter a word as a second argument")
    }

    let number = 0

    for (let i = 0; i < word.length; i++) {
        if (word[i].toLowerCase() === letter.toLowerCase()) {
            number += 1;
        }
    }

    return number
}

//CALL
const result = letterCount("m", "muhammad")
console.log(result)
*/

File: train.py
# M-TASK (PYTHON)
# """
# Shunday function yozing, u string qabul qilsin va string palindrom yani togri oqilganda ham,
# orqasidan oqilganda ham bir hil oqiladigan soz ekanligini aniqlab boolean qiymat qaytarsin.
# MASALAN: palindrom_check("dad") return True;  palindrom_check("son") return False;
# """


# def palindrom_check(word):
#     reverse_word = word[::-1]
#     if word == reverse_word:
#         return True
#     return False


# print(palindrom_check("dam"))


# K-TASK
# '''Shunday function yozing, u string qabul qilsin va string ichidagi eng uzun sozni qaytarsin.
# MASALAN: find_longest("I come from Uzbekistan") return "Uzbekistan"'''


# def find_longest(word):
#     word_array = word.split()
#     longest = ''
#     for i in word_array:
#         if len(i) > len(longest):
#             longest = i
#     return longest


# print(find_longest("MyUzbekistan I come from Uzbekistan"))


# I-TASK
'''Shunday function tuzing, unga string argument pass bolsin. 
Function ushbu agrumentdagi digitlarni yangi stringda return qilsin
MASALAN: get_digits("m14i1t") return qiladi "141"'''


# def get_digits(text):
#     new_string = ''
#     for i in text:
#         if i >= "0" and i <= "9":
#             new_string += i
#     return new_string


# print(get_digits("m14i1t"))

# G-TASK
'''
Shunday function tuzingki unga integerlardan iborat array pass bolsin va 
function bizga osha arrayning eng katta qiymatiga tegishli birinchi indexni qaytarsin.
MASALAN: get_highest_index([5, 21, 12, 21, 8]) return qiladi 1 sonini.
'''


# def get_highest_index(array):
#     high_number = 0
#     high_ind = 0
#     for ind, val in enumerate(array):
#         if val > high_number:
#             high_number = val
#             high_ind = ind
#     return high_ind


# result = get_highest_index([58, 5, 21, 12, 21, 38, 8])
# print(result)
